packages/mobile/src/walletConnect/saga.ts

```diff
@@ -0,0 +1,283 @@
+import { currentAccountSelector } from 'src/web3/selectors'
+
+const TAG = 'WalletConnect/saga'
+
+let client: WalletConnectClient | null = null
+
+export function* acceptSession({ proposal }: AcceptSession) {
+  if (!client) {
+    Logger.debug(TAG + '@acceptSession', 'missing client')
+    return
+  }
+
+  const { nodeDir } = networkConfig
+  const genesis: string = yield call(readGenesisBlockFile, nodeDir)
+  const networkId: number = GenesisBlockUtils.getChainIdFromGenesis(genesis)
+
+  const account: string = yield call(getAccountAddress)
+  const response: SessionTypes.Response = {
+    metadata: {
+      name: 'Valora',
+      description: 'A mobile payments wallet that works worldwide',
+      url: 'https://valoraapp.com',
+      icons: ['https://valoraapp.com/favicon.ico'],
+    },
+    state: {
+      accounts: [`${account}@celo:${networkId}`],
+    },
+  }
+
+  yield call(client.approve.bind(client), { proposal, response })
+}
+
+export function* denySession({ proposal }: DenySession) {
+  if (!client) {
+    Logger.debug(TAG + '@denySession', 'missing client')
+    return
+  }
+
+  yield call(client.reject.bind(client), { reason: 'Session denied by user', proposal })
+}
+
+export function* closeSession({ session }: CloseSession) {
+  if (!client) {
+    Logger.debug(TAG + '@initialiseClient', 'missing client')
+    return
+  }
+
+  yield call(client.disconnect.bind(client), {
+    topic: session.topic,
+    reason: 'Closed by user',
+  })
+  yield put(closeSessionAction(session))
+}
+
+export function* acceptRequest({
+  request: {
+    request: { id, jsonrpc, method, params },
+    topic,
+  },
+}: AcceptRequest) {
+  if (!client) {
+    Logger.debug(TAG + '@initialiseClient', 'missing client')
+    return
+  }
+
+  const account: string = yield select(currentAccountSelector)
+  const wallet: UnlockableWallet = yield call(getWallet)
+
+  let result: any
+
+  if (method === SupportedActions.eth_signTransaction) {
+    yield call(unlockAccount, account)
+    result = (yield call(wallet.signTransaction.bind(wallet), params)) as EncodedTransaction
+  } else {
+    throw new Error('Unsupported action')
+  }
+
+  yield call(client.respond.bind(client), {
+    topic,
+    response: {
+      id,
+      jsonrpc,
+      result,
+    },
+  })
+  yield put(requestFulfilled({ topic, id }))
+
+  navigateBack()
+}
+
+export function* denyRequest({
+  request: {
+    request: { id, jsonrpc },
+    topic,
+  },
+}: DenyRequest) {
+  if (!client) {
+    Logger.debug(TAG + '@initialiseClient', 'missing client')
+    return
+  }
+
+  yield call(client.respond.bind(client), {
+    topic,
+    response: {
+      id,
+      jsonrpc,
+      error: {
+        code: -32000,
+        reason: 'Rejected request',
+      },
+    },
+  })
+  yield put(requestFulfilled({ id, topic }))
+
+  navigateBack()
+}
+
+export function* startWalletConnectChannel() {
+  yield takeLeading(Actions.INITIALISE_CLIENT, watchWalletConnectChannel)
```

Comment:
Can you clarify why the starting of the watcher was split off like this and needs a `takeLeading` instead of just starting it directly from the `walletConnectSaga` function below?

Reply:
+1